openvdb/openvdb/points/PointStatistics.h

```diff
@@ -0,0 +1,609 @@
+///   or define its own numerical limits. This method will throw only if the
+///   templated ValueType does not match the given attribute. This method will
+///   return true if min and max have been set, false otherwise. The function is
+///   deterministic.
+/// @note  The value type of the min/max calculations must match the value type
+///   of the attribute. For vectors and matrices, this results in per component
+///   comparisons. See evalExtents for magnitudes or more custom control.
+/// @param points     the point tree
+/// @param attribute  the attribute to reduce
+/// @param min        the computed min value
+/// @param max        the computed max value
+/// @param filter     a filter to apply to points
+/// @param minTree    if provided, builds a tiled tree of localised min results
+/// @param maxTree    if provided, builds a tiled tree of localised max results
+/// @return true if min and max have been set, false otherwise. Can be false if
+///   no points were processed or if the tree was empty.
+/// @warning if "P" is provided as the attributes, it is evaluated in voxel space
+template <typename ValueT,
+    typename CodecT = UnknownCodec,
+    typename FilterT = NullFilter,
+    typename PointDataTreeT = points::PointDataTree>
+inline bool evalMinMax(const PointDataTreeT& points,
+        const std::string& attribute,
+        ValueT& min,
+        ValueT& max,
+        const FilterT& filter = NullFilter(),
+        typename PointDataTreeT::template ValueConverter<ValueT>::Type* minTree = nullptr,
+        typename PointDataTreeT::template ValueConverter<ValueT>::Type* maxTree = nullptr);
+
+/// @brief Evaluates the average active value of a point attribute.
+/// @details  This function sets the argument average to the average value of a
+///   given point attributes. The ValueType of the attribute must be copy
+///   constructible, support the same type addition operator and the division
+///   operator from a floating point scalar. This method allows for a different
+///   type ResultT to be provided which is used to compute the final result
+///   as well as for intermediate arithmetic. This allows for higher precision
+///   results if overflow or underflow is a concern, as well as computing
+///   floating point results from integer types. This method will throw only if
+///   the templated ValueType does not match the given attribute. This method
+///   will return true if average has been set, false otherwise. The function is
+///   deterministic.
+/// @param points     the point tree
+/// @param attribute  the attribute to reduce
+/// @param average    the computed min value
+/// @param filter     a filter to apply to points
+/// @param averageTree  if provided, builds a tiled tree of localised avg results
+/// @return true if average has been set, false otherwise. Can be false if
+///   no points were processed or if the tree was empty.
+/// @warning if "P" is provided as the attributes, it is evaluated in voxel space
```

Comment:
I agree, we need unique methods/algorithms for position - they don't necessarily need to take a grid though, it could just compute it in index space. But I don't think we should explicitly error here unless we see these methods supporting index space P in the future (in which case we error and add this support later).

Reply:
Average of P is an obvious thing fro computing barycenters, so I'm wondering if there is a more explicity way we could catch this....

Like maybe make it *undefined* and eventually this could switch automatically to evaluatePosition if a voxel-relative attribute is selected?

Reply:
The result for all these methods, if called with 'P', is now described as undefined